DI-1008: assert that tests don't leak threads

bukzor opened the pull request.

bukzor committed:
DI-1008: assert that tests don't leak threads

bukzor committed:
DI-1008: fix thread leaks

kcons commented on src/sentry/testutils/thread_leaks.py:
```diff
@@ -0,0 +1,62 @@
+from psutil._common import pthread
+
+
+def _get_pthreads() -> frozenset[pthread]:
```
Why is it worth involving pthreads here? Is the aim here to find leaks from native extensions? I think it's probably ok to limit ourselves to policing the python runtime until we have cases that require going deeper.

bukzor replied:
You're probably right. I thought I could cover more, fix more at small cost.

kcons commented on src/sentry/testutils/thread_leaks.py:
```diff
@@ -0,0 +1,62 @@
+            continue
+
+        for old_metric, new_metric in zip(old, new):
+            if old_metric > new_metric:  # thread metrics can only grow
```
Not sure I understand this; it starts by comparing TID, which I don't think is guaranteed to be increasing, as they can be reused once a thread exits, iirc.

kcons commented on src/sentry/testutils/thread_leaks.py:
```diff
@@ -0,0 +1,62 @@
+        newstate.assert_no_leak_from(self)
+
+    def assert_no_leak_from(self, oldstate):
+        old_diff, new_diff = _pthread_diff(oldstate.pthreads, self.pthreads)
```
My naive assumption is that we would want to do: 
```
before = frozenset(threading.enumerate())
...
after = frozenset(theading.enumerate())
new_threads = after - before # we probably don't mind if threads go away
assert not new_threads
```

What cases does that not flag that we care abou?

bukzor replied:
duplicate of "Why is it worth involving pthreads here?"

bukzor committed:
fix: shut down singleton producers

bukzor committed:
improve traceback more

bukzor committed:
not all factory "functions" have a __qualname__

bukzor committed:
more stack readability

bukzor committed:
fix(typing)

bukzor committed:
fix: "cannot schedule new futures after shutdown"

bukzor committed:
fix(dlq): use SingletonProducer to handle singleton producers

bukzor committed:
fix thread leak

bukzor committed:
fix thread leak in relay test fixture

bukzor committed:
WIP: trying to fix tests/relay_integration/test_sdk.py::test_recursion_breaker[MONOLITH]

bukzor committed:
implement kill() for MultiplexingTransport

bukzor committed:
supress "your redis package hecka old" warning

this warning needs to be suppressed for anything to work if/when pytest-sentry
is uninstalled

bukzor committed:
avoid problematic `configure_sdk` in test

bukzor committed:
Merge branch 'refs/heads/bukzor/avoid-configure-sdk-in-test'

bukzor committed:
avoid thread leak

bukzor committed:
mypy =.=

untitaker commented on tests/conftest.py:
```diff
@@ -56,6 +58,14 @@ def unclosed_files():
+    with thread_leaks.assert_none():
+        yield
+        # HAX: close all "singleton producers" before checking thread leaks
+        SingletonProducer._shutdown_all()
```
I think it could make sense to make test authors do this:

```
def test_kafka():
    with SingletonProducer.shutdown():
        ...
```

then you don't have to do it here

at the same time, `_shutdown_all` is incredibly cheap if there are no producers, so it makes sense to run it unconditionally

(and i personally like to not have too much indentation in tests

bukzor committed:
allowlist the last few problems, sadface

bukzor committed:
DEBUG: I want to see the flakes, please

bukzor committed:
fix(test): thread leak: trends API thread pool

bukzor committed:
fix(test): thread leak: Snuba search thread pool

bukzor committed:
fix(test): thread leak: organization trace thread pool

bukzor committed:
fix(test): thread leak: trace item attributes ranked

bukzor committed:
fix(test): thread leak: trace meta

bukzor committed:
test thread leak allowlist: KafkaProducer cleanup

bukzor committed:
test thread leak allowlist: Django test server

bukzor committed:
test thread leak allowlist: ingest consumer

bukzor committed:
test thread leak allowlist: profile deobfuscation

bukzor committed:
test thread leak allowlist: relay integration

bukzor committed:
test thread leak allowlist: backup command

bukzor committed:
test thread leak allowlist: Arroyo concurrent execution

bukzor committed:
test thread leak allowlist: preprod background processing

bukzor committed:
test thread leak allowlist: spans flusher database access

bukzor committed:
test thread leak allowlist: taskworker queue threads
